### ISSUE TITLE
Award reputation miners with correct reputation and token rewards

### ISSUE THREAD

**Issue by area** [JoinColony/colonyNetwork#164]
Section 7.8 of the whitepaper describes the amount of reputation that reputation miners get from submitting correct hashes, and needs to be implemented. With the merge of #156, they will only get a fixed amount of reputation in an incorrect skill and a fixed amount of tokens.

A special 'mining' skill must be created in the Common Colony, and the reputation miners awarded this skill.

Approximations will need to be used for the exponential decay in equation (3); so long as it displays the correct quantitative behavior, it will be accepted. The exponential factor in equation (2) can be implemented with a Taylor expansion (see appendix B).

In order to implement equation (3), the time that tokens were staked must be recorded; if a user stakes additional tokens, then the new time calculated should represent the weighted average of the staked tokens. This time does *not* need to be updated when the users are awarded the newly minted tokens as a reward for mining; consider this a benefit of mining (for now, until someone comes up with a reason this is a bad idea).

**Comment by elenadimitrova** [comment]
This issue should cater to solving the remaining `todo`s in `rewardStakers` function [here](https://github.com/JoinColony/colonyNetwork/issues/164)

**Comment by kronosapiens** [comment]
## Continued

After a little more work I believe I can show that any recycled CLNY will never create a total decay greater than the recycled CLNY itself, which means that we can have a fixed inflation rate while allowing for recycling of CLNY.

Also, I have an opinion around the risks of having a variable inflation rate.

## 1: Modeling reputation's "steady state"

First, we should model the way that meta colony reputation accumulates and decays over time, and show that a steady state is indeed reached.

Let `R_n` be the reputation state at period `n`. Let's assume no bootstrapping of reputation, or rather that reputation never increased by more than CLNY in any period. Let `CLNY` be the per-period minting rate, `K` be the decay rate (think ~.2 or so) while `\bar{K}` be equal to `1-K`. Given that the current reputation state `R_n` is the sum of all past decays, we can say that:

```
R_n <= \sum_{i=0}^n CLNY \bar{K}^i
```

With equality if all CLNY is paid out in every period. With this, we are interested in the per-period decay, `KR_n`, which determines the amount of CLNY needed to pay miner rewards. What can we say about `KR_n`?

```
KR_n <= K \sum_{i=0}^n CLNY \bar{K}^i
      = K CLNY \sum_{i=0}^n \bar{K}^i
```

Here we note that the sum is a geometric series, which we solve as per usual:


```
      <= K CLNY 1/(1-\bar{K})
      = K CLNY 1/K
      = CLNY
```

And there we go. As long as no more than CLNY reputation is distributed in every time period, the per-period reputation decay will never exceed CLNY. This is asserted in the WP Sec 7.8 but here's a proof, which is nice.

## Modeling recycled CLNY

The above model was worked out en route to answering the more pressing question: what happens if we "recycle" CLNY (allow people to send CLNY back to the meta colony)? In this case, the amount of CLNY distributed per period is unknown (limited only by the total CLNY in circulation). What effect does this have on the decay rate (and therefore the mining rewards)?

Fortunately, the model looks strikingly similar to the one above. Instead of looking at the reputation state `R_n`, we will concern ourselves with the total decay `D_n` of one injection of recycled CLNY (`rCLNY`). The main difference between these formulate is interpretation: in the first, we were interested in `R_n`, the *current reputation state* as a sum of many mintings of CLNY. Now, we care about `D_n`, which is the *sum of all past decays* of a single recycling.

```
D_n <= \sum_{i=1}^n K rCLNY \bar{K}_i
```

Again, with equality if all of the rCLNY is given out as reputation. What can we say about `D_n`?

```
    = K rCLNY \sum_{i=1}^n \bar{K}_i
    <= K rCLNY 1/(1-\bar{K})
    = K rCLNY 1/K
    = rCLNY
```

Hooray! Even in the case of recycled CLNY, the total increase in decay (and thus miner rewards) will never exceed the amount recycled. Therefore, if we want miner rewards to be 1/4 of the decay, we can simply set aside 1/4 of the rCLNY for mining rewards. Increasing the minting rate should not be necessary.

### Caveat

I believe these results hold (I need to check) only in the case where miner rewards is exclusively a % of reputation decay. If miner rewards are independent of reputation decay, this may not be the case.

To see why, consider a case where some CLNY is set aside for miner rewards and the rest for work. If a fixed amount of CLNY is paid to miners every period regardless of reputation state, then the CLNY set aside for miners could be spent before any CLNY is paid out to workers. Thus, when the CLNY is finally paid out to workers (and is thus incorporated into the reputation state, and thus contributes to decay), there may not be any CLNY left in the mining reward pool to pay for the increase.

## Inflation risks

The last thing to discuss is the risks of inflation, i.e. why we might want to avoid including it if possible.

One of the main problems with inflation is that it devalues existing assets denominated in the inflating currency. If I own a house, and the local currency inflates, the value of my house goes up. No problem there. But if I have $100,000 under my mattress, and the dollar inflates, my savings have lost value. IRL, this discourages people from holding assets denominated in the inflating currency.

For small amounts of inflation, these phenomena are not a big deal, but for large amounts of inflation this can be very destructive. Hence central bankers.

In the case of CLNY, the risk of inflation is that it would "devalue" existing reputation. If we mint 100 CLNY per period, reputation will exist on the order of 10^2. If one day we 10x the minting to 1000 per period, reputation will also go up one order of magnitude. New reputation will quickly eclipse past reputation, leading to a disconnect between the reputation state on-chain and the expectations of the colony participants (who expect their reputation to decay at the rate of K).

I just had this thought, but it seems like the inflation rate and the decay rate are playing similar roles in the meta colony. I wonder how deep that similarity goes... is a variable decay rate essentially equivalent to a variable inflation rate?

## Conclusions

If miner rewards are strictly a function of reputation decay, than variable minting is unnecessary and perhaps undesirable, as it risks destabilizing the reputation state.

**Comment by kronosapiens** [comment]
## Continued II

Ok, I have an idea.

## Bookkeeping for miner rewards

Ideally, we would like a few things:
1. Miner reputation to be about 1/4 (`P`) of the total reputation state.
2. Miner rewards being paid out regardless of other activity in the meta colony
3. A constant minting rate (alternatively, no *need* to mint more CLNY to pay miners)

I think we can have all of these things, if we do the following:

1. Set a minimum reward for mining, to be paid out regardless of the reputation state (`M_{min} = X`)
2. Figure out the miner reward advance `A` as follows: `A = X - RKP`.
3. If `A` is positive (i.e. RKP < X), add `A` to some running total `T` and pay out `X`.
4. If `A` is zero or negative (i.e. RKP > X), subtract `A` from the running total `T` (note that T cannot be less than 0), and then pay out the larger of either `X` or `RKP - T`.

Essentially what this does is ensure that we never pay more than P CLNY out to miners; if we pay the minimum payment `X` in some periods, we make a corresponding deduction in `RKP` in other periods.

This will necessitate an additional storage variable (likely in `ColonyNetwork` somewhere keeping track of a running advance.

**Comment by kronosapiens** [comment]
## Concluding conclusions

TL;DR: everything can work out in the long run without needing a lever for minting CLNY, but a lot of that hinges on the developer incentive mechanism functioning well.

To summarize everything:

1. Under "normal" conditions (ongoing work in the meta colony leading to significant amounts of reputation payouts), we would like reputation miners to hold about 25% of the total reputation. We do this by making miner rewards a function of the reputation state.

2. Another nice thing about this 25% rule is that we can show that, as long as we set aside 25% of incoming CLNY for miner rewards, we will always have enough CLNY to pay miners. In other words, the maximum miner reward as a function of reputation state will never be more than the amount of CLNY which enters the meta colony.

3. Unfortunately, in the case where there is little to no work in the meta colony (an important edge case worth planning for in advance), a simple 25% rule will eventually lead to 0 miner rewards, stopping reputation mining entirely (for all colonies). This is probably the worst outcome.

4. To handle this, we can impose a small "minimum miner reward" to be paid out regardless of the meta colony reputation state. The major risk here is that eventually miners will come to hold 100% of the meta colony reputation. There's no way around that possibility if we want to guarantee some minimum miner reward, although the incorporation of the developer incentive mechanism will make this vanishingly unlikely by adding a second source of automated CLNY payments independent of meta colony work.

5. Unfortunately, by paying out a "minimum miner reward", we lose the guarantee that 25% of incoming CLNY will always be sufficient to meet the miner reward requirements, thus necessitating the need to mint additional CLNY to cover mining rewards.

6. Fortunately, we can get around this via some bookkeeping -- by keeping track of the difference in minimum payouts and reputation-state payouts, we can adjust payments accordingly such that the running total of payouts is never more than 25% of the running total of incoming CLNY, thus obviating the need for a minting lever.

**Comment by kronosapiens** [comment]
So, it seems like some of this analysis, while fun and interesting, was premature -- it doesn't look like a fixed minting rate is a top priority, and we can expect to mint CLNY manually for the time being. Also, I didn't realize that CLNY for miners is minted directly in the `reward` function and is thus independent of other CLNY minting.

So for the first version of the network, I would propose a scheme where the miner reward is `max(min, decay/4)`, with the minimum being determined along the lines of https://github.com/JoinColony/colonyNetwork/issues/356#issuecomment-431276175 and the overall supply being managed manually by the team for the time being.

Once the network matures it would be valuable to circle back around to this issue as the need for a regular and reliable supply of CLNY becomes more important.
